imap/imap.go

```diff
@@ -150,203 +63,176 @@ func Validate(s *models.IMAP) error {
-	cmd, err := imap.Wait(client.UIDSearch(specs...))
+// DeleteEmails will delete emails from the supplied slice of SeqNums
+func (mbox *Mailbox) DeleteEmails(seqs []uint32) error {
+	imapClient, err := mbox.newClient()
 	if err != nil {
-		return &imap.Command{}, fmt.Errorf("uid search failed: %s", err)
+		return err
 	}
-	return cmd, nil
-}
 
-const GenerateBufferSize = 100
+	defer imapClient.Logout()
 
-func (mbox *Mailbox) generateMail(search string, since *time.Time, markAsRead, delete bool) (chan Response, error) {
-	responses := make(chan Response, GenerateBufferSize)
-	client, err := mbox.newClient()
+	seqSet := new(imap.SeqSet)
+	seqSet.AddNum(seqs...)
+
+	item := imap.FormatFlagsOp(imap.AddFlags, true)
+	err = imapClient.Store(seqSet, item, imap.DeletedFlag, nil)
 	if err != nil {
-		close(responses)
-		return responses, fmt.Errorf("failed to create IMAP connection: %s", err)
+		return err
 	}
 
-	go func() {
-		defer func() {
-			client.Close(true)
-			client.Logout(30 * time.Second)
-			close(responses)
-		}()
+	return nil
+}
 
-		var cmd *imap.Command
-		// find all the UIDs
-		cmd, err = findEmails(client, search, since)
-		if err != nil {
-			responses <- Response{Err: err}
-			return
-		}
-		// gotta fetch 'em all
-		getEmails(client, cmd, markAsRead, delete, responses)
-	}()
+// GetUnread will find all unread emails in the folder and return them as a list.
+func (mbox *Mailbox) GetUnread(markAsRead, delete bool) ([]Email, error) {
 
-	return responses, nil
-}
+	imap.CharsetReader = charset.Reader
+	var emails []Email
 
-func getEmails(client Client, cmd *imap.Command, markAsRead, delete bool, responses chan Response) {
-	seq := &imap.SeqSet{}
-	msgCount := 0
-	for _, rsp := range cmd.Data {
-		for _, uid := range rsp.SearchResults() {
-			msgCount++
-			seq.AddNum(uid)
-		}
+	imapClient, err := mbox.newClient()
+	if err != nil {
+		return emails, fmt.Errorf("failed to create IMAP connection: %s", err)
 	}
 
-	if seq.Empty() {
-		return
-	}
+	defer imapClient.Logout()
 
-	fCmd, err := imap.Wait(client.UIDFetch(seq, "INTERNALDATE", "BODY[]", "UID", "RFC822.HEADER"))
+	// Search for unread emails
+	criteria := imap.NewSearchCriteria()
+	criteria.WithoutFlags = []string{"\\Seen"}
+	seqs, err := imapClient.Search(criteria)
 	if err != nil {
-		responses <- Response{Err: fmt.Errorf("unable to perform uid fetch: %s", err)}
-		return
+		return emails, err
 	}
 
-	var email Email
-	for _, msgData := range fCmd.Data {
-		msgFields := msgData.MessageInfo().Attrs
-
-		// make sure is a legit response before we attempt to parse it
-		// deal with unsolicited FETCH responses containing only flags
-		// I'm lookin' at YOU, Gmail!
-		// http://mailman13.u.washington.edu/pipermail/imap-protocol/2014-October/002355.html
-		// http://stackoverflow.com/questions/26262472/gmail-imap-is-sometimes-returning-bad-results-for-fetch
-		if _, ok := msgFields["RFC822.HEADER"]; !ok {
-			continue
-		}
+	if len(seqs) > 0 {
```

Comment:
When possible, I recommend catching and handling conditions early, since that helps avoid long nested blocks.

Here, we can do something like this:

```go
if len(seqs) == 0 {
    return emails, nil
}
// At this point, we know that we have emails we can handle so we can continue as normal.
// The benefit is that we've saved one level of indentation which is cleaner to read.
seqset := new(imap.SeqSet)
...
```

Reply:
Good point, done!